File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis.VisualBasic.Syntax/ArrayRankSpecifierSyntax.cs
using Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax;

namespace Microsoft.CodeAnalysis.VisualBasic.Syntax
{
	public sealed class ArrayRankSpecifierSyntax : VisualBasicSyntaxNode
	{
		public SyntaxToken OpenParenToken => new SyntaxToken(this, ((Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ArrayRankSpecifierSyntax)base.Green)._openParenToken, base.Position, 0);

		public SyntaxTokenList CommaTokens
		{
			get
			{
				GreenNode commaTokens = ((Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ArrayRankSpecifierSyntax)base.Green)._commaTokens;
				return (commaTokens == null) ? default(SyntaxTokenList) : new SyntaxTokenList(this, commaTokens, GetChildPosition(1), GetChildIndex(1));
			}
		}

		public SyntaxToken CloseParenToken => new SyntaxToken(this, ((Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ArrayRankSpecifierSyntax)base.Green)._closeParenToken, GetChildPosition(2), GetChildIndex(2));

		public int Rank => CommaTokens.Count + 1;

		internal ArrayRankSpecifierSyntax(GreenNode green, SyntaxNode parent, int startLocation)
			: base(green, parent, startLocation)
		{
		}

		internal ArrayRankSpecifierSyntax(SyntaxKind kind, DiagnosticInfo[] errors, SyntaxAnnotation[] annotations, PunctuationSyntax openParenToken, GreenNode commaTokens, PunctuationSyntax closeParenToken)
			: this(new Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ArrayRankSpecifierSyntax(kind, errors, annotations, openParenToken, commaTokens, closeParenToken), null, 0)
		{
		}

		public ArrayRankSpecifierSyntax WithOpenParenToken(SyntaxToken openParenToken)
		{
			return Update(openParenToken, CommaTokens, CloseParenToken);
		}

		public ArrayRankSpecifierSyntax WithCommaTokens(SyntaxTokenList commaTokens)
		{
			return Update(OpenParenToken, commaTokens, CloseParenToken);
		}

		public ArrayRankSpecifierSyntax AddCommaTokens(params SyntaxToken[] items)
		{
			return WithCommaTokens(CommaTokens.AddRange(items));
		}

		public ArrayRankSpecifierSyntax WithCloseParenToken(SyntaxToken closeParenToken)
		{
			return Update(OpenParenToken, CommaTokens, closeParenToken);
		}

		internal override SyntaxNode GetCachedSlot(int i)
		{
			return null;
		}

		internal override SyntaxNode GetNodeSlot(int i)
		{
			return null;
		}

		public override TResult Accept<TResult>(VisualBasicSyntaxVisitor<TResult> visitor)
		{
			return visitor.VisitArrayRankSpecifier(this);
		}

		public override void Accept(VisualBasicSyntaxVisitor visitor)
		{
			visitor.VisitArrayRankSpecifier(this);
		}

		public ArrayRankSpecifierSyntax Update(SyntaxToken openParenToken, SyntaxTokenList commaTokens, SyntaxToken closeParenToken)
		{
			if (openParenToken != OpenParenToken || commaTokens != CommaTokens || closeParenToken != CloseParenToken)
			{
				ArrayRankSpecifierSyntax arrayRankSpecifierSyntax = SyntaxFactory.ArrayRankSpecifier(openParenToken, commaTokens, closeParenToken);
				SyntaxAnnotation[] annotations = GetAnnotations();
				if (annotations != null && annotations.Length > 0)
				{
					return SyntaxNodeExtensions.WithAnnotations(arrayRankSpecifierSyntax, annotations);
				}
				return arrayRankSpecifierSyntax;
			}
			return this;
		}
	}
}

File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis/VisualBasic/Syntax/ArrayRankSpecifierSyntax.vb
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.VisualBasic
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
Imports System

Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax
	Public NotInheritable Class ArrayRankSpecifierSyntax
		Inherits Microsoft.CodeAnalysis.VisualBasic.VisualBasicSyntaxNode
		Public ReadOnly Property CloseParenToken As Microsoft.CodeAnalysis.SyntaxToken
			Get
				Return New Microsoft.CodeAnalysis.SyntaxToken(Me, DirectCast(MyBase.Green, Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ArrayRankSpecifierSyntax)._closeParenToken, Me.GetChildPosition(2), MyBase.GetChildIndex(2))
			End Get
		End Property

		Public ReadOnly Property CommaTokens As SyntaxTokenList
			Get
				Dim syntaxTokenLists As SyntaxTokenList
				Dim green As GreenNode = DirectCast(MyBase.Green, Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ArrayRankSpecifierSyntax)._commaTokens
				syntaxTokenLists = If(green Is Nothing, New SyntaxTokenList(), New SyntaxTokenList(Me, green, Me.GetChildPosition(1), MyBase.GetChildIndex(1)))
				Return syntaxTokenLists
			End Get
		End Property

		Public ReadOnly Property OpenParenToken As Microsoft.CodeAnalysis.SyntaxToken
			Get
				Return New Microsoft.CodeAnalysis.SyntaxToken(Me, DirectCast(MyBase.Green, Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ArrayRankSpecifierSyntax)._openParenToken, MyBase.Position, 0)
			End Get
		End Property

		Public ReadOnly Property Rank As Integer
			Get
				Return Me.CommaTokens.Count + 1
			End Get
		End Property

		Friend Sub New(ByVal green As GreenNode, ByVal parent As SyntaxNode, ByVal startLocation As Integer)
			MyBase.New(green, parent, startLocation)
		End Sub

		Friend Sub New(ByVal kind As SyntaxKind, ByVal errors As DiagnosticInfo(), ByVal annotations As SyntaxAnnotation(), ByVal openParenToken As PunctuationSyntax, ByVal commaTokens As GreenNode, ByVal closeParenToken As PunctuationSyntax)
			MyClass.New(New Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ArrayRankSpecifierSyntax(kind, errors, annotations, openParenToken, commaTokens, closeParenToken), Nothing, 0)
		End Sub

		Public Overrides Function Accept(Of TResult)(ByVal visitor As VisualBasicSyntaxVisitor(Of TResult)) As TResult
			Return visitor.VisitArrayRankSpecifier(Me)
		End Function

		Public Overrides Sub Accept(ByVal visitor As Microsoft.CodeAnalysis.VisualBasic.VisualBasicSyntaxVisitor)
			visitor.VisitArrayRankSpecifier(Me)
		End Sub

		Public Function AddCommaTokens(ByVal ParamArray items As Microsoft.CodeAnalysis.SyntaxToken()) As Microsoft.CodeAnalysis.VisualBasic.Syntax.ArrayRankSpecifierSyntax
			Return Me.WithCommaTokens(Me.CommaTokens.AddRange(items))
		End Function

		Friend Overrides Function GetCachedSlot(ByVal i As Integer) As SyntaxNode
			Return Nothing
		End Function

		Friend Overrides Function GetNodeSlot(ByVal i As Integer) As SyntaxNode
			Return Nothing
		End Function

		Public Function Update(ByVal openParenToken As Microsoft.CodeAnalysis.SyntaxToken, ByVal commaTokens As SyntaxTokenList, ByVal closeParenToken As Microsoft.CodeAnalysis.SyntaxToken) As Microsoft.CodeAnalysis.VisualBasic.Syntax.ArrayRankSpecifierSyntax
			Dim arrayRankSpecifierSyntax As Microsoft.CodeAnalysis.VisualBasic.Syntax.ArrayRankSpecifierSyntax
			If (openParenToken <> Me.OpenParenToken OrElse commaTokens <> Me.CommaTokens OrElse closeParenToken <> Me.CloseParenToken) Then
				Dim arrayRankSpecifierSyntax1 As Microsoft.CodeAnalysis.VisualBasic.Syntax.ArrayRankSpecifierSyntax = Microsoft.CodeAnalysis.VisualBasic.SyntaxFactory.ArrayRankSpecifier(openParenToken, commaTokens, closeParenToken)
				Dim annotations As SyntaxAnnotation() = MyBase.GetAnnotations()
				arrayRankSpecifierSyntax = If(annotations Is Nothing OrElse CInt(annotations.Length) <= 0, arrayRankSpecifierSyntax1, arrayRankSpecifierSyntax1.WithAnnotations(annotations))
			Else
				arrayRankSpecifierSyntax = Me
			End If
			Return arrayRankSpecifierSyntax
		End Function

		Public Function WithCloseParenToken(ByVal closeParenToken As Microsoft.CodeAnalysis.SyntaxToken) As Microsoft.CodeAnalysis.VisualBasic.Syntax.ArrayRankSpecifierSyntax
			Return Me.Update(Me.OpenParenToken, Me.CommaTokens, closeParenToken)
		End Function

		Public Function WithCommaTokens(ByVal commaTokens As SyntaxTokenList) As Microsoft.CodeAnalysis.VisualBasic.Syntax.ArrayRankSpecifierSyntax
			Return Me.Update(Me.OpenParenToken, commaTokens, Me.CloseParenToken)
		End Function

		Public Function WithOpenParenToken(ByVal openParenToken As Microsoft.CodeAnalysis.SyntaxToken) As Microsoft.CodeAnalysis.VisualBasic.Syntax.ArrayRankSpecifierSyntax
			Return Me.Update(openParenToken, Me.CommaTokens, Me.CloseParenToken)
		End Function
	End Class
End Namespace